### PR TITLE
feat: Add lost-hours tracking feature

### PR DESCRIPTION
**Pull request by tal**
## Summary

- **Lost hours tracking**: Adds a new feature that allows users to update the cumulative lost hours count in the #lost-hours channel by posting messages like `#lost-hours +=2` or `#lost-hours -=1`
- **Slack API helpers**: New functions for channel operations (`findChannelByName`, `getSlackChannelInfo`, `updateSlackChannelTopic`)
- **Regex patterns**: Dynamic regex generator for matching lost-hours channel mentions with increment/decrement operators

## New OAuth Scopes Required

This feature requires **two new OAuth scopes** to be added to the Slack app:

| Scope | Purpose |
|-------|---------|
| `channels:read` | Required to look up the #lost-hours channel by name using `conversations.list` API |
| `channels:write.topic` | Required to update the channel topic with the new hours count using `conversations.setTopic` API |

### To add these scopes:
1. Go to your Slack app at https://api.slack.com/apps
2. Navigate to **OAuth & Permissions**
3. Under **Bot Token Scopes**, add:
   - `channels:read`
   - `channels:write.topic`
4. Reinstall the app to your workspace to apply the new permissions

## How It Works

1. User posts a message mentioning the #lost-hours channel with an operator (e.g., `#lost-hours +=3`)
2. Bot detects the pattern and looks up the channel by name
3. Parses the current hours from the channel topic (format: "Cumulatively, X hours lost since...")
4. Calculates the new total and updates the channel topic
5. Responds with confirmation (e.g., "Lost hours updated: 1,286 → 1,289 (+3 hours)")

## Test Plan

- [x] Unit tests added for lost-hours message processing
- [x] Tests cover increment, decrement, edge cases, and error handling
- [x] Manual testing with local dev environment using ngrok

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: camille-bot/scripts/start-local-dev.sh [hunk 1]
```diff
@@ -8,7 +8,7 @@ fi
 
 # Start ngrok in the background
 echo "Starting ngrok..."
-ngrok http 8787 > /dev/null &
+ngrok http 8787 --domain secretive-nonfalteringly-alan.ngrok-free.dev > /dev/null &
```
Comment (mergesort) [inline camille-bot/scripts/start-local-dev.sh]: This seems like it shouldn't be committed, right?

File: camille-bot/src/shared/regex/patterns.ts [hunk 2]
```diff
@@ -113,4 +113,25 @@ export function createLeaderboardRegex(botId: string | undefined): RegExp {
...
+ * @param channelId The channel ID to match
+ * @returns A RegExp that matches lost hours modifier commands for that channel
+ *          Capture group 1: operator (+ or -)
+ *          Capture group 2: the number value
+ */
+export function createLostHoursIncrementRegex(channelId: string): RegExp {
+  // Match: <#CHANNELID|name> or <#CHANNELID> or <#CHANNELID|> followed by += or -= and a number
+  // Supports integers and decimals
+  // Note: [^>]* allows zero or more chars after | (Slack sometimes sends empty name)
+  const pattern = `<#${channelId}(?:\\|[^>]*)?>\\s*([+-])=\\s*(\\d+(?:\\.\\d+)?)`;
```
[the remark this answers is not in the document]
Comment (tal) [reply camille-bot/src/shared/regex/patterns.ts]: @codex could you make it surface a validation error in a thread if the users send +/- \d

File: camille-bot/src/lost-hours/lost-hours.ts [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    // Validate the value (must be positive and reasonable)
+    if (rawValue <= 0 || rawValue > 1000) {
+      logger.warn('Invalid lost-hours value', {
+        value: rawValue,
+        operator,
+        userId
+      });
+      return {
+        status: ProcessingStatus.ERROR,
+        response: 'Invalid value. Please use a positive number less than 1000.'
```
[the remark this answers is not in the document]
Comment (tal) [reply camille-bot/src/lost-hours/lost-hours.ts:136]: fixed the test

**Review by mergesort — CHANGES_REQUESTED** [review]
Overall this looks pretty good! I didn't get a chance to run this locally to test it, have you been trying it out? I definitely would like to get this merged in, especially because I know you put a lot of work into it.

I have a few notes and questions that I'd like to see implemented before I can merge this in, but I think it's close.

1. As currently configured I think this will work in every channel, rather than checking that the user is in #found-hours. Is that the intention? I don't think it's necessarily bad, but it may seem unintuitive that this will bump the channel topic of another channel.
2. This changes the topic, but does it also change the description? For historical reasons (Slack's UI changes all the time and is different across platforms) we tend to keep both in sync.
3. Can you add a help command when users run `@camille help` that informs them of the functionality that exists?
4. Can you update the readme to note that this functionality exists the way we do with other features?
5. Could we move `MissingScopeError` into it's own file? I was trying to find it during my review and was surprised to find it tucked into `messaging.ts`.
6. Can we confirm that caching `lostHoursChannel` works, possibly with a test? I'm a little worried about the fact that it fires off a GET request, which could happen every message if we've done this wrong.

Thanks so much for the pull request and the feature @tal!

**Comment by tal** [comment]
> Overall this looks pretty good! I didn't get a chance to run this locally to test it, have you been trying it out? I definitely would like to get this merged in, especially because I know you put a lot of work into it.
> 
> I have a few notes and questions that I'd like to see implemented before I can merge this in, but I think it's close.
> 
> 1. As currently configured I think this will work in every channel, rather than checking that the user is in #found-hours. Is that the intention? I don't think it's necessarily bad, but it may seem unintuitive that this will bump the channel topic of another channel.

That was the intention and i tested that. my thought was someone could type `#lost-hours += 2 whatever they wasted time on` in any channel and it would increment the count. It doesnt keep historical data about who bumped the most or what message they had. Thinking about it now a message in `#lost-hours` with who bumped it and what message they had may be good.

Do you think itd be useful to have the last `N` reasons people stated in the topic as well?

> 2. This changes the topic, but does it also change the description? For historical reasons (Slack's UI changes all the time and is different across platforms) we tend to keep both in sync.

I had no idea the description was also changed. When i contribute to lost hours i dont update the description.

> 3. Can you add a help command when users run `@camille help` that informs them of the functionality that exists?

Sure, didnt know that was even a command

> 4. Can you update the readme to note that this functionality exists the way we do with other features?

Yup

> 5. Could we move `MissingScopeError` into it's own file? I was trying to find it during my review and was surprised to find it tucked into `messaging.ts`.

Sounds good.

> 6. Can we confirm that caching `lostHoursChannel` works, possibly with a test? I'm a little worried about the fact that it fires off a GET request, which could happen every message if we've done this wrong.

I confirmed it myself but while i update this based on the feedback i can again.

**Comment by mergesort** [comment]
> That was the intention and i tested that. my thought was someone could type #lost-hours += 2 whatever they wasted time on in any channel and it would increment the count.

Works for me! This seems like a pretty good user experience.

> Thinking about it now a message in #lost-hours with who bumped it and what message they had may be good.
> Do you think itd be useful to have the last N reasons people stated in the topic as well?

This sounds good in theory but unfortunately channel topics/descriptions have a pretty short character limit so I think keeping the original message and bumping the lost hours count sounds like the best approach.

> I had no idea the description was also changed. When i contribute to lost hours i dont update the description.

Admittedly I'm not in every channel (including #lost-hours), so I just never noticed this discrepancy. Maybe I'll consider moving away from this dichotomy, but I think for now let's just bump both.

> I confirmed it myself but while i update this based on the feedback i can again.

I just realized that rather than checking every time (or if the info isn't cached), we can just do what we do with the `WHERE_DO_I_POST_CHANNEL` parameter. I just grabbed it once from Slack's web interface and then have an environment variable that stores the channel ID — in this case it would be `LOST_HOURS_CHANNEL`. It's a tiny bit more configuration but then we can remove all of the additional code and querying to cache the channel ID.

@tal How do those ideas feel to you?

File: camille-bot/scripts/start-local-dev.sh (hunk 1 above)
Comment (tal) [reply camille-bot/scripts/start-local-dev.sh]: alredy fixed in a subsequent commit. i immediatly noticed it after i posted

**Comment by tal** [comment]
> I just realized that rather than checking every time (or if the info isn't cached), we can just do what we do with the WHERE_DO_I_POST_CHANNEL parameter. I just grabbed it once from Slack's web interface and then have an environment variable that stores the channel ID — in this case it would be LOST_HOURS_CHANNEL. It's a tiny bit more configuration but then we can remove all of the additional code and querying to cache the channel ID.

¯\_(ツ)_/¯ it's already built and since i needed the additional scope for changing the topic felt like it was worth it.

**Comment by mergesort** [comment]
Even though it's already built I'd prefer to do it the simple way rather than adding new code for handling scopes and additional complexity. Building upon this pattern will make the bot easier to maintain in the future, without having the an additional layer of work for any channel-related queries -- especially since the channel ids are never going to change.

**Comment by DollyParton247** [comment]
Hello  everyone I'm new here

**Review by mergesort — APPROVED** [review]

**Comment by mergesort** [comment]
Not sure if I fixed that merge conflict correctly but let's give it a shot. 😂
